packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1243 @@
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private focusType: TableFocusType = TableFocusType.none;
+    private headerActionIndex = -1;
+    private rowIndex = -1;
+    private cellContentIndex = -1;
+    private columnIndex = -1;
+    private inNavigationMode = true;
+    private readonly tableNotifier: Notifier;
+    private readonly virtualizerNotifier: Notifier;
+    private visibleRowNotifiers: Notifier[] = [];
+
+    public constructor(
+        private readonly table: Table<TData>,
+        private readonly virtualizer: Virtualizer<TData>
+    ) {
+        table.addEventListener('keydown', e => this.onCaptureKeyDown(e), {
+            capture: true
+        });
+        table.addEventListener('keydown', e => this.onKeyDown(e));
+        table.addEventListener('focusin', e => this.handleFocus(e));
+        this.tableNotifier = Observable.getNotifier(this.table);
+        this.tableNotifier.subscribe(this, 'rowElements');
+        this.virtualizerNotifier = Observable.getNotifier(this.virtualizer);
+        this.virtualizerNotifier.subscribe(this, 'visibleItems');
+    }
+
+    public resetFocusState(): void {
+        this.focusType = TableFocusType.none;
+        const activeElement = this.getActiveElement();
+        if (activeElement && this.isInTable(activeElement)) {
+            this.setDefaultFocus();
+        }
+    }
+
+    public connect(): void {
+        this.table.viewport.addEventListener('keydown', e => this.onViewportKeyDown(e));
+        this.table.viewport.addEventListener('cell-action-menu-blur', e => this.onCellActionMenuBlur(e));
+    }
+
+    public handleChange(source: unknown, args: unknown): void {
+        let focusRow = false;
+        if (source === this.virtualizer && args === 'visibleItems') {
+            focusRow = true;
+        } else if (source === this.table && args === 'rowElements') {
+            for (const notifier of this.visibleRowNotifiers) {
+                notifier.unsubscribe(this);
+            }
+            this.visibleRowNotifiers = [];
+            for (const visibleRow of this.table.rowElements) {
+                const rowNotifier = Observable.getNotifier(visibleRow);
+                rowNotifier.subscribe(this, 'dataIndex');
+                if (visibleRow.dataIndex === this.rowIndex) {
+                    focusRow = true;
+                }
+            }
+        } else if (args === 'dataIndex') {
+            const dataIndex = (source as TableRow | TableGroupRow).dataIndex;
+            if (dataIndex === this.rowIndex) {
+                focusRow = true;
+            }
+        }
+
+        if (focusRow) {
```

Comment:
This block doesn't necessarily focus a row, correct? It could focus a cell inside the row. If so, I would prefer this local be called something like `focusRowOrCell`.

Reply:
Renamed.